### PR TITLE
feat: convert project to be expo managed

### LINKED ISSUE #225: Adopt Expo Prebuild
Switch from a RN bare work flow to [Continuous Native Generation](https://docs.expo.dev/workflow/continuous-native-generation/) (see [Adopt Prebuild](https://docs.expo.dev/guides/adopting-prebuild/)).

To Do:
- [x] Determine any modifications made to native code (can create a new bare react native project and compare android folders for diffs)
- [x] For any modification made, determine dependencies required the native code changes.
- [x] Determine if there are expo alternatives for these libraries OR if we should create an [Expo Plugin](https://docs.expo.dev/config-plugins/plugins-and-mods/). If there is an alternative library that can be used consult @achou11 or @ErikSin to determine if we should switch libraries
- [x] [Follow Expo instruction to set up app.config.js](https://docs.expo.dev/build-reference/variants/#using-production-variant) to configure the [development, preview, and production builds](https://docs.expo.dev/build/eas-json/#common-use-cases). For now the build configurations will be the same, but create the 3 categories in the app config for later configuration
- [x] Figure out the work needed to get [nodejs-mobile-react-native](https://nodejs-mobile.github.io/) working with Expo
    - based on the [documented installation instructions](https://nodejs-mobile.github.io/docs/guide/guide-react-native/getting-started), there maybe be some additional work needed so that the nodejs-assets/ directory is placed into the correct place in the generated android/ directory.

### LINKED ISSUE #268: `npm run android` shouldn't require `npm start` to run
You can't run `npm run android` without a Metro server running (in other words, by running `npm start`).

This causes confusion for new developers and we might want to fix this.

### PR DESCRIPTION
**Pull request by CDFN**
This PR converts project to be expo managed meaning we don't include `android/` and `ios/` folders anymore.
It also adds configuration for Expo services.

Closes #225 
Closes #268 (now `npm run android` bundles and installs app to device and runs server via expo)

### CODE REVIEW

File: patches/nodejs-mobile-react-native+18.17.7+004+fix-delete-incorrect-prebuilds-gradle-step.patch [hunk 1]
```diff
@@ -1,12 +1,38 @@
...
+diff --git a/node_modules/nodejs-mobile-react-native/scripts/ios-sign-native-modules.sh b/node_modules/nodejs-mobile-react-native/scripts/ios-sign-native-modules.sh
+index 81c033b..96bf544 100644
+--- a/node_modules/nodejs-mobile-react-native/scripts/ios-sign-native-modules.sh
++++ b/node_modules/nodejs-mobile-react-native/scripts/ios-sign-native-modules.sh
+@@ -8,7 +8,7 @@ if [ -f ./.xcode.env.local ]; then
+   source "./.xcode.env.local";
+ fi
+ 
+-DESIRED_NODE_VERSION="18"
++DESIRED_NODE_VERSION="$(node -p "process.versions.node.split('.')[0]")"
```
Comment (achou11) [inline patches/nodejs-mobile-react-native+18.17.7+004+fix-delete-incorrect-prebuilds-gradle-step.patch:35]: what's the context for this patch? can this be split into a separate patch using patch-package's `--append` option?

https://github.com/ds300/patch-package/?tab=readme-ov-file#creating-multiple-patches-for-the-same-package

File: patches/nodejs-mobile-react-native+18.17.7+005+disable-node-version-check.patch
Comment (achou11) [inline patches/nodejs-mobile-react-native+18.17.7+005+disable-node-version-check.patch:1]: is removing this patch intentional?

**Review by achou11 — CHANGES_REQUESTED** [review]
Ran out of time today so not an exhaustive review - looking forward to testing this out though!

File: app.config.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  {development: '.dev', production: '', test: '.test'}[
+    process.env.APP_VARIANT
+  ] ?? '';
+
+const NAME =
+  {development: ' (DEV)', production: '', test: ' (TEST)'}[
+    process.env.APP_VARIANT
+  ] ?? '';
+
+module.exports = ({config}) => ({
```
[the remark this answers is not in the document]
Comment (CDFN) [reply app.config.js:11]: Ah, I remember there was a way to do it but couldn't find it in google when worked on that file! 😅 Yeah, converting to TS is not so good, as transpiling expo configs is acting little funky when I tested it, especially when you import something (let alone importing other TS file 😁) so I would just stick to JS as it seems good enough.

File: app.json [hunk 3]
```diff
@@ -1,14 +1,70 @@
...
+      "bundleIdentifier": "com.comapeo"
+    },
+    "extra": {
+      "eas": {
+        "projectId": "2d5b8137-12ec-45aa-9c23-56b6a1c522b7"
       }
-    ]
-  ]
+    },
+    "owner": "digidem"
```
[the remark this answers is not in the document]
Comment (CDFN) [reply app.json:68]: You can change it anytime (and project id accordingly), so it's nothing permanent. That's used for Expo services configuration.

File: patches/nodejs-mobile-react-native+18.17.7+004+fix-delete-incorrect-prebuilds-gradle-step.patch (hunk 1 above)
Comment (CDFN) [reply patches/nodejs-mobile-react-native+18.17.7+004+fix-delete-incorrect-prebuilds-gradle-step.patch:35]: Forgot to mention it in readme. Basically this patch just asusmes your node version is desired one, as app is ran with node 20 while embeded nodejs for running backend is 18 which caused issues. This is just android patch ported to support ios too.

**Review by achou11 — CHANGES_REQUESTED** [review]
Having issues getting this to run on my device, but that could be an issue on my end

**Comment by achou11** [comment]
Adding here to track: I had to pass the `--localhost` flag for `expo start` (default is `host`) in order to get the app running on a _physical_ device. this seems like an Expo bug but haven't found anything reported just yet. Ideally this should be working on LAN 🤔

File: app.json [hunk 4]
```diff
@@ -1,14 +1,70 @@
...
+        {
+          "isIosBackgroundLocationEnabled": true,
+          "isAndroidBackgroundLocationEnabled": true
+        }
+      ],
+      [
+        "expo-camera",
+        {
+          "cameraPermission": "Allow $(PRODUCT_NAME) to access your camera",
+          "microphonePermission": "Allow $(PRODUCT_NAME) to access your microphone",
```
[the remark this answers is not in the document]
Comment (CDFN) [reply app.json:38]: I don't think it supports translations whatsoever. It's used only for iOS though, so for now I guess it's not a big deal. We've just added it for sake of having complete configuration 😅

File: src/frontend/contexts/IntlContext.tsx
[the remark this answers is not in the document]
Comment (CDFN) [reply src/frontend/contexts/IntlContext.tsx:1]: I know, I was very suprised too it doesn't work. However for some reason the data would be undefined whenever I tried to debug this code as it caused errors. It is called once only so I guess should be fine to leave it as it is? Didn't encounter any issues in other places in project so I think it's just some weird edge case.

File: package.json [hunk 5]
```diff
@@ -2,24 +2,26 @@
...
+  "main": "index.js",
   "scripts": {
-    "build:backend": "node ./scripts/build-backend.mjs",
     "android": "npm run build:backend && npm run android-no-backend-rebuild",
-    "android-no-backend-rebuild": "react-native run-android",
-    "ios": "react-native run-ios",
-    "prestart": "npm run build:translations && npm run build:intl-polyfills",
-    "start": "react-native start",
+    "android-no-backend-rebuild": "expo run:android",
+    "ios": "expo run:ios && npm run android-no-backend-rebuild",
```
[the remark this answers is not in the document]
Comment (CDFN) [reply package.json:9]: absolutely no, not sure how it ended up here, thanks for catching 😅

File: package.json [hunk 6]
```diff
@@ -2,24 +2,26 @@
...
+    "prestart": "npm run build:translations && npm run build:intl-polyfills && npm run build:backend",
```
[the remark this answers is not in the document]
Comment (CDFN) [reply package.json:10]: Okay, that seems fair. Currently it'll look like that:
- `postinstall` triggers rebuild of all autogenerated files (translations, polyfills, backend),
- `prestart` triggers generation of frontend stuff only (translations, polyfills),
- `android` triggers rebuild of frontend + backend + enables android app,
- `ios` triggers rebuild of frontend + backend + enables ios app,
- `android-no-backend-rebuild`, as name says, just rebuilds frontend and enables android app,
- `ios-no-backend-rebuild` is analogous to android task,
- `build:autogenerated` is just alias for generating translations, polyfills, backend

LMK if we should add some setup notes to readme, as this autogenerated setup is a little bit tricky to understand at first glance.

File: metro.config.js
[the remark this answers is not in the document]
Comment (bohdanprog) [reply metro.config.js:1]: you're right, we can remove that patch.

File: metro.config.js [hunk 7]
```diff
@@ -1,26 +1,29 @@
...
-    // For https://github.com/kristerkari/react-native-svg-transformer
+  const {transformer, resolver} = config;
+
+  config.transformer = {
+    ...transformer,
     babelTransformerPath: require.resolve('react-native-svg-transformer'),
-  },
-  resolver: {
-    // For nodejs-mobile
-    blockList: [/nodejs-assets\/.*/],
```
[the remark this answers is not in the document]
Comment (bohdanprog) [reply metro.config.js:19]: yeah, thanks, I missed it

File: metro.config.js
[the remark this answers is not in the document]
Comment (CDFN) [reply metro.config.js:1]: changed it to just return an object

File: patches/nodejs-mobile-react-native+18.17.7+005+disable-node-version-check.patch
Comment (CDFN) [reply patches/nodejs-mobile-react-native+18.17.7+005+disable-node-version-check.patch:1]: file restored + merged with ios version check path

**Review by achou11 — CHANGES_REQUESTED** [review]
Noticed that the input elements are different due to the prebuild generating some base styles at the native level in `android/app/src/main/res/values/styles.xml` (see screenshots below):

```xml
<resources xmlns:tools="http://schemas.android.com/tools">
  <style name="AppTheme" parent="Theme.AppCompat.Light.NoActionBar">
    <item name="android:textColor">@android:color/black</item>
    <item name="android:editTextStyle">@style/ResetEditText</item>
    <item name="android:editTextBackground">@drawable/rn_edit_text_material</item>
    <item name="colorPrimary">@color/colorPrimary</item>
    <item name="colorPrimaryDark">@color/colorPrimaryDark</item>
  </style>
  <style name="ResetEditText" parent="@android:style/Widget.EditText">
    <item name="android:padding">0dp</item>
    <item name="android:textColorHint">#c8c8c8</item>
    <item name="android:textColor">@android:color/black</item>
  </style>
  <style name="Theme.App.SplashScreen" parent="AppTheme">
    <item name="android:windowBackground">@drawable/splashscreen</item>
  </style>
</resources>
```

Not entirely against this happening since it kind of acts like a style reset but it does mean that we need to go through the app to see what needs adjusting. Luckily there aren't too many places with inputs, I think...

More generally, do you know where the docs for configuring this are?

---

On my Pixel 6, Android 14...

Before:

image

After:

image

File: babel.config.js [hunk 8]
```diff
@@ -1,8 +1,12 @@
...
-  ],
+module.exports = function (api) {
+  api.cache(true);
+  return {
+    //https://github.com/babel/babel/discussions/13676#discussioncomment-1183149
+    compact: true,
+    presets: ['babel-preset-expo'],
+    plugins: [
+      'transform-inline-environment-variables',
+      'react-native-reanimated/plugin',
```
[the remark this answers is not in the document]
Comment (bohdanprog) [reply babel.config.js:9]: restored.

File: patches/nodejs-mobile-react-native+18.17.7+004+fix-delete-incorrect-prebuilds-gradle-step.patch
[the remark this answers is not in the document]
Comment (bohdanprog) [reply patches/nodejs-mobile-react-native+18.17.7+004+fix-delete-incorrect-prebuilds-gradle-step.patch:1]: We removed both path-packages and created a new one which includes changes for both platforms.

[the remark this answers is not in the document]
Comment (CDFN) [reply patches/nodejs-mobile-react-native+18.17.7+004+fix-delete-incorrect-prebuilds-gradle-step.patch:1]: That was my fault, I got lost when merging 2 patches into one when cleaning up. Sorry! Should be sorted now.

**Comment by bohdanprog** [comment]
> Noticed that the input elements are different due to the prebuild generating some base styles at the native level in `android/app/src/main/res/values/styles.xml` (see screenshots below):
> 
> ```
> <resources xmlns:tools="http://schemas.android.com/tools">
>   <style name="AppTheme" parent="Theme.AppCompat.Light.NoActionBar">
>     <item name="android:textColor">@android:color/black</item>
>     <item name="android:editTextStyle">@style/ResetEditText</item>
>     <item name="android:editTextBackground">@drawable/rn_edit_text_material</item>
>     <item name="colorPrimary">@color/colorPrimary</item>
>     <item name="colorPrimaryDark">@color/colorPrimaryDark</item>
>   </style>
>   <style name="ResetEditText" parent="@android:style/Widget.EditText">
>     <item name="android:padding">0dp</item>
>     <item name="android:textColorHint">#c8c8c8</item>
>     <item name="android:textColor">@android:color/black</item>
>   </style>
>   <style name="Theme.App.SplashScreen" parent="AppTheme">
>     <item name="android:windowBackground">@drawable/splashscreen</item>
>   </style>
> </resources>
> ```
> 
> Not entirely against this happening since it kind of acts like a style reset but it does mean that we need to go through the app to see what needs adjusting. Luckily there aren't too many places with inputs, I think...
> 
> More generally, do you know where the docs for configuring this are?
> 
> On my Pixel 6, Android 14...
> 
> Before:
> 
> image
> After:
> 
> image

Hey, right, we missed that, I added an expo-config-plugin that removed those styles.

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (achou11) [reply package.json]: more of a personal preference, but not entirely sure if we need `postinstall` to do all of the work related to autogenerated files. at the moment, the ones in question are only needed for when you want to run the app (or tests once we have more of them) so i'd think that doing that work in `prestart` and before `android` would be okay. i like having `npm install` being as fast as possible, with it mostly being used to deal with setting up the dev environment (not necessarily app code). again, not adamant about this as it's just my personal preference.

otherwise your breakdown seems reasonable to me!

> LMK if we should add some setup notes to readme, as this autogenerated setup is a little bit tricky to understand at first glance.

Think it's okay to not document this for now since things are still changing a bit. The long-term plan is to have a contributors/development-specific doc that details this stuff, which I think would be easier for someone like myself to do :)

**Comment by achou11** [comment]
@CDFN  @bohdanprog heads up that i cleaned up the custom expo config plugin implementation in https://github.com/digidem/comapeo-mobile/pull/300/commits/14be5c8a42112f2992e5ce3aca288f6aa2f7d9fe. figured it would be easier to do it myself than have more back and forth since it was the last thing i wanted to address before approving the PR :)

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (achou11) [reply package.json]: after some more thinking, think we should remove the `build:autogenerated` script and remove its usage from `postinstall`. think it will make sense to revisit all the scripts setup in a separate PR, but introducing it here unnecessarily increases CI time (more specifically running the `build:backend` command when it's not needed for the CI tasks)

**Review by achou11 — APPROVED** [review]
Overall changes are now working well, terrific work! 👍 Any smaller issues are/can be follow-up PRs
